src/browser/BrowserAction.cpp

```diff
@@ -0,0 +1,534 @@
+            return handleGetDatabaseHash(json, action);
+        } else if (action.compare("associate", Qt::CaseSensitive) == 0) {
+            return handleAssociate(json, action);
+        } else if (action.compare("test-associate", Qt::CaseSensitive) == 0) {
+            return  handleTestAssociate(json, action);
+        } else if (action.compare("get-logins", Qt::CaseSensitive) == 0) {
+            return handleGetLogins(json, action);
+        } else if (action.compare("generate-password", Qt::CaseSensitive) == 0) {
+            return handleGeneratePassword(json, action);
+        } else if (action.compare("set-login", Qt::CaseSensitive) == 0) {
+            return handleSetLogin(json, action);
+        } else if (action.compare("lock-database", Qt::CaseSensitive) == 0) {
+            return handleLockDatabase(json, action);
+        }
+    }
+    return getErrorReply(action, ERROR_KEEPASS_INCORRECT_ACTION);
+}
+
+const QJsonObject BrowserAction::handleChangePublicKeys(const QJsonObject& json, const QString& action)
+{
+    const QString nonce = json.value("nonce").toString();
+    const QString clientPublicKey = json.value("publicKey").toString();
+
+    if (!clientPublicKey.isEmpty()) {
+        QMutexLocker locker(&m_mutex);
+        m_associated = false;
+        unsigned char pk[crypto_box_PUBLICKEYBYTES];
+        unsigned char sk[crypto_box_SECRETKEYBYTES];
+        crypto_box_keypair(pk, sk);
+
+        const QString publicKey = getBase64FromKey(pk, crypto_box_PUBLICKEYBYTES);
+        const QString secretKey = getBase64FromKey(sk, crypto_box_SECRETKEYBYTES);
+        m_clientPublicKey = clientPublicKey;
+        m_publicKey = publicKey;
+        m_secretKey = secretKey;
+
+        QJsonObject response;
+        response["action"] = action;
+        response["publicKey"] = publicKey;
+        response["nonce"] = nonce;
+        response["version"] = KEEPASSX_VERSION;
+        response["success"] = "true";
+
+        return response;
+    }
+    return getErrorReply(action, ERROR_KEEPASS_CLIENT_PUBLIC_KEY_NOT_RECEIVED);
+}
+
+const QJsonObject BrowserAction::handleGetDatabaseHash(const QJsonObject& json, const QString& action)
+{
+    const QString hash = getDataBaseHash();
+    const QString nonce = json.value("nonce").toString();
+    const QString encrypted = json.value("message").toString();
+    const QJsonObject decrypted = decryptMessage(encrypted, nonce, action);
+
+    if (!decrypted.isEmpty()) {
+        QString command = decrypted.value("action").toString();
+        if (!hash.isEmpty() && command.compare("get-databasehash", Qt::CaseSensitive) == 0) {
+            QJsonObject message;
+            message["hash"] = hash;
+            message["version"] = KEEPASSX_VERSION;
+
+            QJsonObject response;
+            response["action"] = action;
+            response["message"] = encryptMessage(message, nonce);
+            response["nonce"] = nonce;
```

Comment:
In addition, you could perhaps always save the last nonce and use `sodium_compare()`, to check that the new nonce is bigger to prevent replay attacks.

Reply:
I'm ashamed because I have totally missed this. Will fix it. Needs changes to the browser extension side too.

Reply:
On both sides, check that a new nonce is old nonce + 1. If it's not, reject the message and reset the session.